![AIMS Logo](https://www.uni-bielefeld.de/fakultaeten/mathematik/ag/hofmanova/daad-aims/aims-senegal.png)

****

# <center> <b> <span style="color:orange;"> Python Proficiency for Scientific Computing and Data Science (PyPro-SCiDaS)  </span> </b></center>

### <center> <b> <span style="color:green;">An Initiation to Programming using Python (Init2Py) </span> </b></center>
    

# Parfait TOLEFO 

****
# <center> <b> <span style="color:blue;">Assignment </span> </b></center>

****
### Instructions:
You should solve these problems using only:
- Standard Python data types (e.g., str, list, dict, int).
- Built-in and user-defined functions.
- Basic flow control (if/else, for loops, while loops).
- Standard library modules such as math
****

## Section 1: Python Fundamentals

### Exercise 1
Compute the following expressions for a natural number $n$ chosen by the user:

a) $$\sum_{\substack{i=0 \\ i \text{ is odd}}}^{n-1} (i + 3)$$


In [64]:
# Your code here
try:
    n=int(input("Enter a naturel number: "))
    result=0
    for i in range(n):
        if i%2!=0:
            result+=(3+i)
    print("The result is: ",result)
except:
    print("Error, you enter an invalid number !")

Enter a naturel number: 5
The result is:  10


b) $$\prod_{\substack{i=0 \\ i \text{ is even}}}^{n-1} (i + 3)$$

In [1]:
# Your code here

try:
    n=int(input("Enter a naturel number: "))
    result=1
    for i in range(n):
        if i%2==0:
            result*=(3+i)
    print("The result is: ",result)
except:
    print("Error, you enter an invalid number !")

Enter a naturel number: 5
The result is:  105


### Exercise 2
a) Write python code that computes the sum of the square of the first $n$ natural numbers. One version with a "for loop" and another version with a "while loop".

In [66]:
# Part a: For loop version
n=int(input("Enter a naturel number: "))
result=0
# We suppose that first n natural number are the number from 1,2,3..n
for i in range(n+1):
    result+=i**2

print(result)

Enter a naturel number: 5
55


In [67]:
# Part a: While loop version
n=int(input("Enter a naturel number: "))
result=0

while n:
    result+=n**2
    n-=1
print(result)

Enter a naturel number: 5
55


b) Write python code that prints, for a natural number $M$, the smallest natural number $n$ such that $1^2 + 2^2 + \cdots + n^2 \geq M$. One version with a "for loop" and another version with a "while loop".

In [68]:
# Part b: While loop version
def smallest_square(n):
    """
    This function compute the sum of square of of all number <= n
    """
    res=0
    while n:
        res+=n**2
        n-=1
    return res

__name__=='__main__'

try:
    M=int(input("Enter a naturel number: "))
    result=0; i=0

    while 1:
        if(smallest_square(i)==M):
            print(i)
            break
        elif(smallest_square(i)<M and smallest_square(1+i)>M):
            print(i)
            break
        i+=1
except:
    print("Error, you enter an invalid number !")

Enter a naturel number: 5
2


In [69]:
# Part b: For loop version

def smallest_square(n):
    """
    This function compute the sum of square of of all number <= n
    """
    res=0
    n+=1
    for i in range(n):
        res+=i**2
    return res
    

__name__=='__main__'

M=int(input("Enter a naturel number: "))
result=0

for i in range(M+1):
    
    if(smallest_square(i)==M):
        print(i)
        break
    elif(smallest_square(i)<M and smallest_square(1+i)>M):
        print(i)
        break

Enter a naturel number: 5
2


### Exercise 3
Write a python script that asks the user for the values of $x$ and $y$ and uses these inputs to compute the numerical value of the expression:

$$\frac{\sqrt{x^2+y}}{\frac{1}{x} + 2\sin y}$$

In [73]:
# Your code here

import math

try:
    x=int(input("Enter a number (x): "))
    y=int(input("Enter a number (y): "))

    num = math.sqrt((x**2)+y)
    deno=(1/x)+(2*math.sin(y))
    result=num/deno
    print("The result is: ",result)
except:
    print("Error, you enter an invalid number !")

Enter a number (x): 1
Enter a number (y): 1
The result is:  0.527112989542452


### Exercise 4

A store charges XOF $12$ per item if fewer than $10$ items are purchased. If the number of items is between $10$ and $99$ (inclusive), the price is XOF $10$ per item. For $100$ items or more, the price drops to XOF $7$ per item.

a) Write a function called `Cost` that takes as input the number of items $n$ and returns the total cost.


In [80]:
# Your code here
def Cost(n):
    """
        This takes as input the number of items $n$ and returns the total cost
    """
    if n < 10:
        return 12*n
    elif n==10 or n<=99:
        return 10*n
    else:
        return n*7


700

b) Test the function `Cost` for $n = 50 \text{ and } 255$.

In [81]:
# Your code here
print("For n=50, Cost = ",Cost(50))
print("For n=255, Cost = ",Cost(255))

For n=50, Cost =  500
For n=255, Cost =  1785


### Exercise 5
For a fixed real number $x$ and a natural number $n$, we can define recursively $x^n$ using the relations:

$$x^0 = 1 \quad \text{and} \quad x^{n+1} = x \cdot x^n$$

Write a function `power(x,n)` that implements the above recursion.

In [95]:
# Your code here
def power(x,n):
    if n==0:
        return 1
    else:
        return x*power(x,n-1)

power(2,8)

256

### Exercise 6

a) Write two functions: `arithmetic_mean`, which takes two numbers and returns their [**arithmetic mean**](https://en.wikipedia.org/wiki/Arithmetic_mean)
, and `geometric_mean`, which takes two numbers and returns their [**geometric mean**](https://en.wikipedia.org/wiki/Geometric_mean). Then write a function `avggeom` that takes two input numbers and, for 10 iterations, repeatedly computes both their arithmetic mean and geometric mean, using the results from the previous iteration as the new inputs.


In [3]:
# Your code here

from math import sqrt
def arithmetic_mean(a,b):
    return (a+b)/2

def geometric_mean(a,b):
    return sqrt(a*b)

def avggeom(a,b):
    for i in range(10):
        geo=geometric_mean(a,b)
        ari=arithmetic_mean(a,b)
        print(f"Tour {i+1} => {geo,ari}")
        a=geo;b=ari
avggeom(5,6)

Tour 1 => (5.477225575051661, 5.5)
Tour 2 => (5.4886009750012015, 5.488612787525831)
Tour 3 => (5.488606881260338, 5.488606881263516)
Tour 4 => (5.488606881261927, 5.488606881261927)
Tour 5 => (5.488606881261927, 5.488606881261927)
Tour 6 => (5.488606881261927, 5.488606881261927)
Tour 7 => (5.488606881261927, 5.488606881261927)
Tour 8 => (5.488606881261927, 5.488606881261927)
Tour 9 => (5.488606881261927, 5.488606881261927)
Tour 10 => (5.488606881261927, 5.488606881261927)


b) Write a function called `distancefn` that computes the [**Euclidean distance**](https://en.wikipedia.org/wiki/Euclidean_distance) between two points $(x_1, y_1)$ and $(x_2, y_2)$ in the plane. Use it in another program that computes the perimeter and the area of a triangle, given three points in the plane (each given as coordinates).

 


In [8]:
# Your code here
from math import sqrt

def distancefn(p,q):
    return sqrt((p[0]-q[0])**2+(p[1]-q[1])**2)

distancefn((5,6),(4,3))  

3.1622776601683795

In [13]:
def triangle(a,b,c):
    ab=distancefn(a,b)
    bc=distancefn(b,c)
    ac=distancefn(a,c)
    perimeter=ab+bc+ac
    cords=[ab,ac,bc];cords.sort()
    print(cords)
    area=(cords[0]*cords[1])/2
    
    return {"Perimeter":perimeter,"Area":area}
    
triangle((5,6),(4,3),(6,4))

[2.23606797749979, 2.23606797749979, 3.1622776601683795]


{'Perimeter': 7.63441361516796, 'Area': 2.5000000000000004}

### Exercise 7

a) Let `L=[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]`. Use list comprehension to produce a list of the gaps (differences) between consecutive entries in `L`. Then find the maximum gap size and the percentage of gaps that have size $2$.

In [139]:
# Your code here
L=[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]
l=L
my_l=[L[i+1]-L[i] for i in range(len(L)-1)]
max_gap=max(my_l)

for i in my_l:
    if my_l.count(i)==2:
        
        avg=(2/len((my_l)))*100
        break

print(f"""
The max gap = {max_gap}

The average of gap with size 2 ({i})= {avg}%

    """)


The max gap = 6

The average of gap with size 2 (6)= 14.285714285714285%

    


b) i) Using list comprehensions, write a function `ListComp` that takes a list of strings as input and returns three objects:

- a list that consists of the strings with their first characters removed
- a list of the lengths of the strings
- a list that consists of only those strings that are at least 5 characters long


In [125]:
# Your code here
def ListComp(L):
    return [string[1::] for string in L],[len(word) for word in L], [big_str for big_str in L if len(big_str)>=5]

ii) Test your function with the list: `["hello", "world", "python", "programming", "AI", "code"]`

In [131]:
# Your code here
L=["hello", "world", "python", "programming", "AI", "code"]
without_first_characters, lenghts, longue_str =ListComp(L)

print(without_first_characters,lenghts, longue_str, sep='\n\n')

['ello', 'orld', 'ython', 'rogramming', 'I', 'ode']

[5, 5, 6, 11, 2, 4]

['hello', 'world', 'python', 'programming']


### Exercise 8

a) Write a program that repeatedly asks the user to enter product names and prices. Store all of these in a dictionary whose keys are the product names and whose values are the prices. When the user is done entering products and prices, allow them to repeatedly enter a product name and print the corresponding price or a message if the product is not in the dictionary.

In [ ]:
# Your code here
products={}
stop=input('Enter "Product Name" and Price as the format: Product_Name:Price, when done enter "DONE"\n\n')

while str(stop).upper()!='DONE':
    if ':' in stop and stop.split(':')[1].isnumeric():
        stop=stop.split(':')
        try:
            trash=products[stop[0]]
            print(f"The product {stop[0]} was update")
            products[stop[0]]=stop[1]
            stop=input("")
        except:
            products[stop[0]]=stop[1]
            stop=input()
    else:
        stop=input("Wrong input !\n")

stop=input('Enter "Product Name" to print the Product Price, when done enter "DONE"\n\n')
while str(stop).upper()!='DONE':
    try:
        print("The price of ",stop,"is ",products[stop])
        stop=input()
    except:
        print("The product ",stop," doesn't exist !")
        stop=input()
    

Enter "Product Name" and Price as the format: Product_Name:Price, when done enter "DONE"

banana:2585
Tomato:785
Jeans:965
Done
Enter "Product Name" to print the Product Price, when done enter "DONE"

Jeans
The price of  Jeans is  965
Banana
The product  Banana  doesn't exist !
banana
The price of  banana is  2585


b) Using the dictionary created above, allow the user to enter an amount and print out a list of the products whose price is less than that amount.

In [169]:
# Your code here
mount=float(input("Enter an amount: "))
for i in products.items():
    if float(float(i[1])<mount):
        print(i[0] ,'=>', i[1])

Enter an amount: 42520
Banana => 20525
Tomato => 475


c) Using the dictionary created above, evaluate the most expensive product.

In [174]:
# Your code here
mounts=products.values()
max_mount=max(mounts)
for i in products.items():
    if i[1]==max_mount:
        print("The expensive product is ",i[0]," => ",max_mount)
        break

The expensive product is  Clothes  =>  55854


d) Using the dictionary created above, calculate the average price of all products.

In [179]:
# Your code here
mounts=products.values()
mounts=[float(i) for i in mounts]
t_mount=sum(mounts)
avg=t_mount/len(mounts)
print("The average of price is: ",avg)

The average of price is:  33109.5


e) Using the dictionary created above, update the prices of multiple products at once.

In [191]:
# Your code here
stop=input('Enter "Product Name" to update with the New Price "Product_name:Price", when done enter "DONE"\n\n')

while stop!='DONE':
    if ':' in stop and stop.split(':')[1].isnumeric():
        stop=stop.split(':')
        try:
            trash=products[stop[0]]
            print(f"The product {stop[0]} was update")
            products[stop[0]]=stop[1]
            stop=input("")
        except:
            print(f"The product {stop[0]} doesn't exist !",)
            stop=input("")
    else:
        stop=input("Wrong input !\n")

Enter "Product Name" to update with the New Price "Product_name:Price", when done enter "DONE"

Sample
Wrong input !
Sample:852
The product Sample doesn't exist !
Tomato:8524
The product Tomato was update
Banana:75842
The product Banana was update
Jeans:7854
The product Jeans was update
DONE


### Exercise 9

Use string methods and attributes to solve the following:

a) Write a function called `word_counter` that reads in the file `example.txt` and estimates the total number of words in the file. Assume words are separated by spaces.

In [195]:
!cat example.txt

There are 4 word


In [198]:
# Your code here

file="example.txt"
f=open(file,'r')
lines=f.readlines()
f.close()
word_count=0
for i in lines:
    j=i.split(' ')
    word_count+=len(j)
print(f"\nThe estimate word in the file is {word_count}\n")


The estimate word in the file is 4



b) Write a function called `UpLow` that reads the content of `example.txt` and then returns a dictionary containing the total number of uppercase letters and the total number of lowercase letters.

In [206]:
# Your code here
def UpLow(file="example.txt"):
    """
    This funtion return in a dictionnary the number of lower and upper charecter, the numbers are a lower letters
    """
    f=open(file,'r')
    lines=f.readlines()
    f.close()
    lines_count=len(lines)
    Up_count=0
    Low_count=0
    for i in lines:
        for j in i:
            if j==j.lower() and not j.isnumeric():
                Low_count+=1 
            elif j==j.upper() and not j.isnumeric():
                Up_count+=1
            elif j.isnumeric():
                print(j," is a number")
            else:
                print(j," is a special char")
    return {"Upper_ount":Up_count,"Lower_count":Low_count}
UpLow()

4  is a number


{'Upper_ount': 1, 'Lower_count': 15}

c) Write a function called `count_word` that reads in `example.txt` and returns how many times the `target_word` ("the") appears in the text. Take note that the count must be case-insensitive (e.g., "The", "the", "THE" all count as one) and that it must only count whole word matches, not partial matches (e.g., "them" should not count if the target is "the").

In [215]:
# Your code here
def count_word(target_word="the",file="example.txt"):
    target_word_bak=target_word
    target_word=target_word.upper()
    f=open(file,'r')
    lines=f.readlines()
    f.close()
    word_count=0
    for line in lines:
        line=line.split(' ')
        for i in line:
            if target_word == i.upper():
                word_count+=1
    return (target_word_bak,word_count)
count_word("Are")

('Are', 1)

d) Write a function called `swap_and_clean` that reads the content of the file `example.txt`. The function should process the text by converting all uppercase letters to lowercase and all lowercase letters to uppercase. It should also remove all characters that are not letters, such as numbers and punctuation, while keeping spaces intact. After performing these modifications, the function should write the resulting text to a new file called `output1.txt`.

In [287]:
# Your code here
def swap_and_clean(file_in="example.txt",file_out="output1.txt"):
    f=open(file,'r')
    content=f.read()
    f.close()
    content=list(content)
    to_del=[]
    for i in range(len(content)):
        if content[i].isalpha():
            if content[i].islower():
                content[i]=content[i].upper()
            elif content[i].isupper():
                content[i]=content[i].lower()
            else:
                print("Unknow char => ",content[i])
        else:
            to_del.append(i)
    
    new_content=''
    for i in range(len(content)):
        if i not in to_del:
            new_content+=content[i]
        else:
            continue
    f=open(file_out,"w")
    f.write(new_content)
    f.close()
    return new_content

swap_and_clean()

'tHEREAREWORDmADAMeLLE'

e) Write a function called `palindromes` that reads the content of `example.txt`. The function should first split the text into individual words. For each word, it should be cleaned by converting it to lowercase and removing any leading or trailing punctuation. After cleaning, the function should check whether the word is a [**palindrome**](https://en.wikipedia.org/wiki/Palindrome). A palindrome is a word that reads the same forwards and backwards eg "madam". If a word is a palindrome, the function should print the original, uncleaned version of that word to the console.

In [257]:
# Your code here
def palindromes(file="example.txt"):
    f=open(file,'r')
    content=f.read()
    f.close()
    content=content.replace('\n',' ')
    content=content.split(' ')
    clear_content={}
    for word in content:
        clear_word=''
        for j in word:
            if str(j).isalpha():
                clear_word+=j
        if clear_word:
            clear_content[word]=clear_word.lower()
    for item in clear_content.items():
        if item[1] == item[1][::-1]:
            print(f"{item[0]} is a palindrome")

palindromes()

Madam is a palindrome
Elle is a palindrome


f) Write a function called `extract_and_format_numbers` that reads the file `example.txt` line by line. The function should identify any numbers that appear in the text, such as `100000000`. For each number found, it should format the number by inserting commas in the appropriate places (for example, converting `100000000` to `100,000,000`). Each formatted number should then be written on a separate line in a new file called `output2.txt`.

In [2]:
# Your code here
def extract_and_format_numbers(file_in="example.txt",file_out="output2.txt"):
    f=open(file_in,'r')
    line=f.readline()
    words=[]
    while line:
        words.extend(line[0:len(line)-1].split(' '))
        line=f.readline()
    f.close()

    f=open(file_out,'w')
    for num in words:
        if num.isnumeric():
            num=int(num)
            f_num=f"{num:,}"+'\n'
            f.write(f_num)
        
extract_and_format_numbers()
       

In [3]:
!cat output2.txt

4
100,000,520
10,000,041
100,000


### Exercise 10

The [**digital root**](https://en.wikipedia.org/wiki/Digital_root) of a positive number $n$ is obtained as follows: 
- Add up the digits $n$ to get a new number.
- Add up the digits of that to get another new number. Keep doing this until you get a number that has only one digit. That number is the digital root.

For example, if $n = 45893$, we add up the digits to get $$4 + 5 + 8 + 9 + 3 = 29.$$ We then add up
the digits of $29$ to get $$2 + 9 = 11.$$ We then add up the digits of $11$ to get $$1 + 1 = 2.$$ Since 2 has
only one digit, 2 is our digital root.

Write the Python code of the function `DigRoot` that returns the digital root of an integer $n$ given as input.



In [4]:
# Your code here
def DigRoot(n):
    while True:
        list_n=list(str(n))
        list_n=[int(i) for i in list_n]
        root_n=sum(list_n)
        if len(str(root_n))==1:
            return root_n
        else:
            return DigRoot(root_n)

DigRoot(45893)

2

## Section 2: Mathematical Applications

### Exercise 1
Compute the first 30 terms of the sequence defined by $2x_{n+2} - x_{n+1} - 6x_n = 0$ sequence, given any two initial values $x_0$ and $x_1$.

In [21]:
# Your code here
def computing(x_0,x_1):
    terms=[x_0,x_1]
    for i in range(28):
        terms.append((1/2)*((terms[-1])+(6*terms[-2])))
    return terms

computing(1,0)

[1,
 0,
 3.0,
 1.5,
 9.75,
 9.375,
 33.9375,
 45.09375,
 124.359375,
 197.4609375,
 471.80859375,
 828.287109375,
 1829.5693359375,
 3399.64599609375,
 7188.531005859375,
 13793.203491210938,
 28462.194763183594,
 55610.70785522461,
 113191.93821716309,
 223428.09267425537,
 451289.86098861694,
 895929.2085170746,
 1801834.1872243881,
 3588704.719163418,
 7199854.921254873,
 14366041.61811769,
 28782585.572823465,
 57489417.6407648,
 115092465.5388528,
 230014485.6917208]

### Exercise 2

Let $f$ be a continuous function defined on the closed interval $[a, b]$ such that $f(a)$ and $f(b)$ have opposite signs. By the [**Intermediate Value Theorem**](https://en.wikipedia.org/wiki/Intermediate_value_theorem), there exists a root $x_0 \in (a, b)$ for which $f(x_0) = 0$.

The [**bisection method**](https://fiveable.me/numerical-analysis-i/unit-3/bisection-method-algorithm/study-guide/5LfBknp8yfSe5lse) is an iterative method for approximating the root of such a function. At each step, it divides the interval in half and selects the subinterval where the sign change occurs, repeating this process until the interval width is smaller than a given tolerance ($\epsilon$).

a) Write a function named `bisection(f, a, b, epsilon=1e-2)` that:

- Returns a tuple $(\bar{x}, f(\bar{x}))$, where $\bar{x}$ is an approximation of the root such that
$$
|\bar{x} - x_0| \le \epsilon
$$
if $f(a)\cdot f(b) < 0$.  

- Prints ``Bisection method failed'' if $f(a)\cdot f(b) \ge 0$.


In [39]:
# Your code here
def bisection(f, a, b, epsilon=1e-2):
    f_a,f_b=f(a),f(b)
    if f_a * f_b > 0:
        print("Bisection method failed")
        return None
    elif f_a==0:
        return a,f_a
    elif f(b)==0:
        return b,f_b

    while abs((b - a) / 2) > epsilon:
        half = (a + b) / 2
        f_h=f(half)
        if f_h == 0:
            return half, f_h

        if f(a) * f_h < 0:
            b=half
        else:
            a=half
    return (a + b) / 2, f((a + b) / 2)


b) Test your function `bisection` for $ f(x) = x^3 -3x - 5$, $a=-10$ and $b=10$

In [41]:
# Your code here
f=lambda x: x**3-(3*x)-5
bisection(f,-10,10)

(2.275390625, -0.04555877298116684)

### Exercise 3
An irrational decimal fraction, [**Champernowne constant**](https://en.wikipedia.org/wiki/Champernowne_constant), is created by concatenating the positive integers: 0.12345678910<font color='red'>1</font>112131415161718192021...

It can be seen that the 12th digit of the fractional part is <font color='red'>1</font>. If $d_n$ represents the nth digit of the fractional part, find the value of the following expression.

$$d_1 × d_{10} × d_{100} × d_{1000} × d_{10000} × d_{100000} × d_{1000000}$$

In [50]:
def champernowne_digit(n):
    k, digits, count = 1, 1, 9
    while n > digits * count:
        n -= digits * count
        k += count
        digits += 1
        count *= 10
    k += (n-1)//digits
    return int(str(k)[(n-1)%digits])

indices = [1,10,100,1000,10000,100000,1000000]
prod = 1
for i in indices:
    prod *= champernowne_digit(i)
print(prod)

210


## Section 3: Advanced Programming

### Exercise 1

a) Write a Python function named `factorial2` that takes a non-negative integer $n$ as input and returns $n!$ (the factorial of $n$).

In [13]:
# Your code here

def factorial2(n=0):
    if n <=0:
        return 1
    else:
        return n*factorial2(n-1)
factorial2(10)

3628800

b) i) Write a Python function named `QuadEqu` that solves a quadratic equation of the form:
$$ax^2 + bx + c = 0$$

Your function should:
- Take three real numbers $a$, $b$, and $c$ as inputs, with $a$ defaulting to 1.
- Return the real solutions of the equation (if any) in a suitable format.

In [56]:
# Your code here
from math import sqrt

def QuadEqu(c,b,a=1):  
    """
    This function solves a quadratic equation by taken a,b,c in the following order: c,b,a.
    'a=1' by default so, given to values means will be assign to b and c.
    """
    a=float(a);b=float(b);c=float(c)
    discr=(b**2)-(4*c*a)
    if discr <0:
        return None
    if discr==0:
        sol=((-1)*b)/(2*a)
        return sol
    else:
        sol_1=(((-1)*b)-sqrt(discr))/(2*a)
        sol_2=(((-1)*b)+sqrt(discr))/(2*a)
        return sol_1, sol_2

ii) Test the function `QuadEqu` for:
  - $a=1, b=2, c=1$
  - $a=2, b=2, c=40$
  - $a=2, b=3, c=-5$

In [64]:
# Your code here

print(QuadEqu(1,2,1))
print(QuadEqu(40,2,2))
print(QuadEqu(-5,3,2))

-1.0
None
(-2.5, 1.0)


c) Write a Python program named `fact2nddeg` that repeatedly offers the user a menu with three options:
- Compute a factorial using factorial2
- Solve a quadratic equation using QuadEqu
- Exit the program

The program should:
- Ask the user to select an option.
- Request the necessary input values for the chosen option.
- Call the corresponding function and display the result.
- Repeat the process until the user chooses to exit.

In [120]:
# Your code here
# fact2nddeg.py

def menu():
    choice=input("""
    Enter a choice to continue...
    
    1- Compute a factoriel
    2- Solve a quadratic equation
    3- Exit
    """)
    return choice

__name__='__main__'

choice = menu()

while choice !=3:
    if choice.strip()=='3':
        print("Bye")
        break
    elif choice.strip()=='1':
        try:
            n=int(input("Enter an interger: "))
            fact=factorial2(n)
            print("Factorial of ",n," = ",fact)
        except:
            print("Wrong input !")
    elif choice.strip()=='2':
        vals=input("""Enter a,b and c separed by space (a=1 by default): """).split(' ')
        if len(vals)==2:
            res=QuadEqu(vals[1],vals[0])
        elif len(vals)==3:
            res=QuadEqu(vals[2],vals[1],vals[0])
        else:
            print("Wrong input !")
        if res==None:
            print("\nNo solution of this equation")
        else:
            print("\nThe solution ",res)
    else:
        print("Wrong input !")
    choice = menu()


    Enter a choice to continue...
    
    1- Compute a factoriel
    2- Solve a quadratic equation
    3- Exit
    &
Wrong input !

    Enter a choice to continue...
    
    1- Compute a factoriel
    2- Solve a quadratic equation
    3- Exit
    &éééééééééééééééééee1
Wrong input !

    Enter a choice to continue...
    
    1- Compute a factoriel
    2- Solve a quadratic equation
    3- Exit
    3
Bye


### Exercise 2
This exercise builds the essential mathematical tools used in public-key cryptography, like [**Rivest–Shamir–Adleman (RSA)**](https://en.wikipedia.org/wiki/RSA_cryptosystem). Security relies on the unique properties of prime numbers and the extreme difficulty of factoring very large numbers.

The Core Principle: Prime Factoring is Hard
   - **Easy**: Multiplying two large prime numbers ($p$ and $q$) to get a huge composite number ($N=p \cdot q$) is very fast for computers.
   - **Hard**: Given only the product $N$, finding the original two prime factors, $p$ and $q$, takes an enormous amount of time. The time required grows exponentially with the size of $N$. This difficulty is what guarantees security.

a) Write a function called `number_of_factors` that takes a positive integer $n$ and returns how many factors the number has together with a list of its factors.

In [71]:
# Your code here

def number_of_factors(n):
    
    if n <= 0:
        return None
    
    factors=[]
    for i in range(1, n):
        for j in range(1, n):
            if i*j==n:
                factors.append((i,j))
    
    return len(factors), factors

number_of_factors(20)

(4, [(2, 10), (4, 5), (5, 4), (10, 2)])

b) Write a Python function called `is_prime` that takes a positive integer, $n$, as input and returns the boolean True if $n$ is a prime number and False, otherwise.

Note : A prime number (or a prime) is a natural number greater than 1 and that has no positive divisors other than 1 and itself.

In [42]:
# Your code here

def is_prime(n):
    for i in range(2,n):
        if n%i==0:
            return False
    return True

is_prime(13)

True

c) Cryptographic systems need a large pool of primes. Write a function called `primes(k)` that returns a list of the first k prime numbers. (Set the default value of $k$ to 100).

In [83]:
# Your code here
def primes(k=100):
    list_prime=[]
    i=2
    while True:
        if is_prime(i):
            list_prime.append(i)
            if len(list_prime)==k:
                break
        i+=1
    return list_prime

primes(7)
        

[2, 3, 5, 7, 11, 13, 17]

d) Modify the `primes` function to accept an optional argument called start (default value is 2). The function should now return the first $k$ prime numbers that are greater than or equal to start. This simulates searching for primes in a specific high-security range.

In [61]:
# Your code here

def primes(k=100,start=2):
    list_prime=[]
    while True:
        if is_prime(start):
            list_prime.append(start)
            if len(list_prime)==k:
                break
        start+=1
    return list_prime

primes(5,100)

[101, 103, 107, 109, 113]

e) The [**Fundamental theorem of arithmetic**](https://en.wikipedia.org/wiki/Fundamental_theorem_of_arithmetic) states that every integer $n$ is a prime number or can be uniquely written as a product of prime powers (e.g., $12=2^2 \cdot 3^1$. We are interested in finding the exponent of a specific prime in such a factorization. Write a function `V(n, p)` that takes a positive integer $n$ and a prime $p$, and returns the largest integer $r$ such that $p^r$ divides $n$. This value, $r$, is known as the [**p-adic valuation**](https://en.wikipedia.org/wiki/P-adic_valuation).

In [93]:
# Your code here

def exp_count(x, i):
    """
    This function take a number n and check how many time we
    can have i like x=i^count*...
    """
    count = 0
    while x % i == 0:
        x //= i
        count += 1
    return count

        
def V(n,p):
    if  not is_prime(p):
        print(f"{p} is not a prime number")
        return None
    if (n<=0):
        print(f"The min value of n should be 2")
        return None
    prod=1
    i=2
    n1=n
    while prod<n1:
        if is_prime(i):
            exp=exp_count(n,i)
            n=n//(i**exp)
            prod*=(i**exp)
            if(i==p):
                return exp
        i+=1
    return None

print(V(-12,3))
print(V(12,2))          

The min value of n should be 2
None
2


f) We would like now to compare the prime-factor structure of two numbers, $n$ and $m$, that is, we would like to measure how different two numbers are in terms of the powers of the primes in their factorizations. Write a function `D(n, m)` that takes two positive integers, $n$ and $m$, and returns a total score $S_1$ calculated as follows:

 $$ S_1 = \sum_{p\; prime} \Bigl| \nu_p(n) - \nu_p(m)\Bigr| $$

Note that `D(14,24)=4`.



In [131]:
def V_2(n):
    
    """
    This function take an interger N and return a dictionnary that content the primer factor as key and their
    exponet as value
    """
    primes_dic={}
    if (n<=0):
        #print(f"The min value of n should be 2")
        return None
    prod=1
    i=2
    n1=n
    while prod<n1:
        if is_prime(i):
            exp=exp_count(n,i)
            n=n//(i**exp)
            prod*=(i**exp)
            primes_dic[i]=exp
        i+=1
    return primes_dic
print(V_2(24))
print(V_2(14))

{2: 3, 3: 1}
{2: 1, 3: 0, 5: 0, 7: 1}


In [132]:
# Your code here
def D(n,m):
    primes_n=V_2(n)
    primes_m=V_2(m)
    s_1=0
    try:
        if len(primes_m)>len(primes_n):
            for i in primes_m.keys():
                try:
                    s_1+=abs(primes_n[i]-primes_m[i])
                except:
                    s_1+=abs(primes_m[i])
        else:
            for i in primes_n.keys():
                try:
                    s_1+=abs(primes_n[i]-primes_m[i])
                except:
                    s_1+=abs(primes_n[i])
    except:
            pass
    return s_1

D(0,1)

0

g) The final problem requires you to analyze the average behavior of the `D(n,m)` function by computing a large, iterated sum.  Hence, write a function `P` that takes a positive integer `N` and returns the sum $S_2$ given by:

$$S_2  = \sum_{n=1}^{N}\sum_{m=1}^{N}D(n,m).$$

Note that $P(10)=210$ and $P(10^2) = 37018$.


In [137]:
# Your code here
def P(N):
    s_2=0
    for i in range(N+1):
        for j in range(N+1):
            try:
                s_2+=D(i,j)
            except:
                continue
    return s_2


print(P(10))
print(P(100))

210
37018


### Exercise 3

Imagine a population of bacteria that grows at a rate proportional to its current size, but with limited food supply. Suppose its growth is modeled by

$$\dfrac{dy}{dt}=0.5y$$

with initial population $y(0)=2$.

a) Explain in words what this equation means about how the population changes over time.

**Your explanation here**

_The constant 0.50 corresponds to the growth rate: it tells us that at any given moment, the population increases by 50% of its current size per unit of time._

b) [**Euler’s method**](https://en.wikipedia.org/wiki/Euler_method) is a numerical technique to approximate solutions of ordinary differential equations (ODEs) when an exact solution is hard (or impossible) to find.

Suppose we have an ODE:
$$\dfrac{dy}{dt} = f(t, y), \quad y(t_0)=y_0$$

The idea is:

- Start from an initial point $(t_0, y_0)$
- Move forward in small steps of size $h$
- At each step, estimate the new value of $y$ using the slope $f(t,y)$ at the current point

The update formula is thus given by:
$$y_{n+1} = y_n + h \cdot f(t_n,y_n)$$
$$t_{n+1} = t_n + h$$

The pseudocode is given as follows:

c) i) Write $f(t, y) = 0.5y$ as a lambda function.

In [142]:
f = lambda t,y: 0.5 * y

ii) Now, use Euler’s method with a step size $h=0.1$ to approximate the population at $t=0.3$. Strictly use lists.

In [157]:
t0 = 0
y0 = 2
h = 0.1
n = 3 
t = [t0]
y = [y0]

for i in range(n):
    y_next = y[i] + h * f(t[i], y[i])
    t_next = t[i] + h
    y.append(y_next)
    t.append(t_next)

print(y[-1])


2.3152500000000003


d) Compare your result at $t=0.3$ with the exact solution $y(t) = 2\exp^{0.5t}$, that is, compute the absolute error.

In [161]:
import math

def y(t):
    return 2 * math.exp(0.5 * t)

print(y(0.3))


2.323668485456566


my answer is close to the real enswer: 2.3152500000000003~=2.323668485456566

e) Compare your answers for $y(0.3)$ with $h=0.1$ and $h=0.05$. Which one is closer to the exact solution $y(0.3) = 2\exp^{0.15}$ ? Reflect on this.

In [162]:
t0 = 0
y0 = 2
h = 0.05
n = 3 
t = [t0]
y = [y0]

for i in range(n):
    y_next = y[i] + h * f(t[i], y[i])
    t_next = t[i] + h
    y.append(y_next)
    t.append(t_next)

print(y[-1])

2.1537812499999998


In [163]:
import math

def y(t):
    return 2 * math.exp(0.5 * t)

print(y(0.15))


2.155768301769263


my answer is close to the real enswer: 2.1537812499999998 ~= 2.155768301769263


### Exercise 4

Determine how many Sundays fell on the first day of a month during the twentieth century, that is, from 1 January 1901 to 31 December 2000.

The following information might be useful (though you may also verify or research independently):

- 1 January 1900 was a Monday.
- The number of days in each month is as follows:
    - September, April, June, and November each have 30 days.
    - All other months have 31 days, except February, which has 28 days in a common year and 29 days in a leap year.
- Leap year rules:
    - A year is a leap year if it is divisible by 4.
    - However, if the year is a century (i.e., divisible by 100), it is only a leap year if it is also divisible by 400.

In [164]:
# Your code here